HW0/P4/P4.py

```diff
@@ -66,11 +130,87 @@
     # B = ?
     # C = ?
 
+    # make B
+    B = np.zeros([6,6])
+    B = np.matrix(B)
+
+    B[0,0] = bx
+    B[1,1] = by
+    B[2,2] = bz
+    B[3,3] = bvx
+    B[4,4] = bvy
+    B[5,5] = bvz
+
+    # make C
+    C = np.matrix(np.zeros([3,6]))
+    C[0,0] = rx
+    C[1,1] = ry
+    C[2,2] = rz
+
     # Initial conditions for s0 and Sigma0
+    s = np.matrix([0, 0, 2, 15, 3.5, 4.0]).T
+
+    Sigma = np.matrix(np.zeros([6,6]))
+    for i in range(6):
+        Sigma[i,i] = 1
+    Sigma = 0.01 * Sigma
+
+    # will store in S again
+    S = np.zeros([6,K])
+    S = np.matrix(S)
+    S[:,0] = s
+
+
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    # Defining functions
+
+    # Eq(2)
+    def predictS(sk):
+        return (A * sk) + a
+
+    # Eq(3)
+    def predictSig(sigk):
+        return ((A * sigk * A.T) + (B * B.T)).I
+
+    # Eq(4)
+    def updateSig(approx_sig):
+        return (approx_sig + (C.T * C)).I
+
+    # Eq(5)
+    def updateS(sigk, approx_sig, approx_s, mk):
+        return sigk * ((approx_sig * approx_s) + (C.T * mk))
+
+
+    # putting steps all together
+    i = 1
+    while i < K:
+        # predict
+        approx_s = predictS(s)
+        approx_Sigma = predictSig(Sigma)
+
+        # update
+        Sigma = updateSig(approx_Sigma)
+        approx_Xk = np.matrix(approx_Xk)
+        curr_mk = robot_mk[i,:].T
+
+        s = updateS(Sigma, approx_Sigma, approx_s, curr_mk)
+
+        # record
+        S[:,i] = s
+
+        # step
+        i = i + 1
+
+    # graph it
+    S = S.T
+    S = np.array(S)
+    x_coords = S[:,0]
+    y_coords = S[:,1]
+    z_coords = S[:,2]
+
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    ax.plot(x_coords, y_coords, z_coords,
+        '-r', label='Filtered trajectory')
 
     # Show the plot
     ax.legend()
```

Comment:
Your code is very clean and easily understandable by an external reader. However, as the calculation of S is very compact, it might be a good place to add a comment to help your reader (especially a reader who did not do the Homework and will struggle more to follow your logic).